### PR TITLE
fix: dont try to coerce list for regex match

### LINKED ISSUE #11622: Internal error when regex operator `~` is used with `List`s (SQLancer)
### Describe the bug

The following query has an invalid operation, so it should return a planning error.
However, it's returning an internal error to indicate a potential bug
```
> select [1,2] ~ [3];
Internal error: invalid operator for nested.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

Found by SQLancer https://github.com/apache/datafusion/issues/11030

### PR DESCRIPTION
**Pull request by tshauck**
## Which issue does this PR close?

Closes #11622

## Rationale for this change

Regex match's args will try to be coerced from Lists, but that shouldn't be possible.

List coercion was added in https://github.com/apache/datafusion/pull/10445, but it's not clear to me why.

## What changes are included in this PR?

Remove list coercion, add tests

## Are these changes tested?

yes

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/sqllogictest/test_files/regexp.slt [hunk 1]
```diff
@@ -325,6 +325,11 @@ SELECT 'foo\nbar\nbaz' ~ 'bar';
 ----
 true
 
+statement error
+Error during planning: Cannot infer common argument type for regex operation List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata
```
Comment (tshauck) [inline datafusion/sqllogictest/test_files/regexp.slt:329]: A bit verbose, but tries to prove the error happens during planning.

**Comment by jayzhan211** [comment]
I guess we don't need list coercion for `Like` too

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by alamb** [comment]
cc @b41sh  (as the author of https://github.com/apache/datafusion/pull/10445) in case you have extra context you would like to share

**Comment by tshauck** [comment]
> I guess we don't need list coercion for Like too

+1, I'll wait a day or so to hear if it's there for a reason for it, otherwise I'll pull it out too.

**Comment by alamb** [comment]
Thanks again @tshauck  and @jayzhan211

**Pull request merged**